### PR TITLE
Blend shapes functionality

### PR DESCRIPTION
**Pull request by Sovogal**
Added blend shape functionality if the user specifies script define LATIOS_MECANIM_EXPERIMENTAL_BLENDSHAPES
Includes mecanim transition fix

### CODE REVIEW

File: Packages/com.latios.latiosframework/Kinemation/Authoring/BakingSystems/MeshDeformDataSmartBlobberSystem.cs [hunk 1]
```diff
@@ -152,10 +152,10 @@ protected unsafe override void OnUpdate()
...
                     }
                     builder.blendShapeBufferSize = verticesCount;
-                    var cpuBuffer                = new NativeArray<BlendShapeVertexDisplacement>(gpuBuffer.count * 10,
-                                                                                                 Allocator.Persistent,
-                                                                                                 NativeArrayOptions.UninitializedMemory);
-                    blendShapeRequests[index]       = UnityEngine.Rendering.AsyncGPUReadback.RequestIntoNativeArray(ref cpuBuffer, gpuBuffer);
+                    var cpuBuffer =
+                        new NativeArray<BlendShapeVertexDisplacement>((int)verticesCount, Allocator.Persistent, NativeArrayOptions.UninitializedMemory);
+                    blendShapeRequests[index]       = UnityEngine.Rendering.AsyncGPUReadback.RequestIntoNativeArray(ref cpuBuffer, gpuBuffer, cpuBuffer.Length, 0);
+
```
Comment (Dreaming381) [inline Packages/com.latios.latiosframework/Kinemation/Authoring/BakingSystems/MeshDeformDataSmartBlobberSystem.cs]: How is this working for you? This change is reverting the bugfix I made that fixed blend shape rendering. Granted, I now am realizing that there's a bug where multiplying by 10 is wrong and should be a divide or perhaps I should leave the array as uint (the 10 comes from the number of uints that fit in a BlendShapeVertexDisplacement). But the RequestIntoNativeArray with 4 arguments is just flat out broken in Unity at the moment and has been confirmed by QA.

File: Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs [hunk 2]
```diff
@@ -458,11 +459,172 @@ internal static bool Approximately(float a, in float b)
...
+                    continue;
+
+                ref var state = ref controllerBlob.layers[clipWeight.layerIndex].states[clipWeight.stateIndex];
+                var time = state.isLooping ? clip.LoopToClipTime(clipWeight.motionTime) : math.min(clipWeight.motionTime, clip.duration);
+                var stateSpeed = state.speedMultiplierParameterIndex != -1 ?
+                                 parameters[state.speedMultiplierParameterIndex].floatParam * state.speed :
+                                 state.speed;
+                var speedModifiedDeltaTime = deltaTime * stateSpeed;
+                var deltaTransform = TransformQvvs.identity;
+                var hasLooped = state.isLooping && time - deltaTime < 0f;
```
Comment (Dreaming381) [inline Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs]: Shouldn't this be speedModifiedDeltaTime here and shouldn't there be a check for the reverse direction too?

File: Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs [hunk 3]
```diff
@@ -458,11 +459,172 @@ internal static bool Approximately(float a, in float b)
...
+                //If the clip has looped, get a sample of the end of the clip to incorporate it into the delta
+                if (hasLooped)
+                {
+                    deltaTransform = clip.SampleBone(0, time);
+                    var previousClipSample = clip.SampleBone(0, time - speedModifiedDeltaTime);
+
+                    var sampleEnd = math.select(clip.duration, -clip.duration, stateSpeed < 0f);
+                    var endClipSample = clip.SampleBone(0, sampleEnd);
+
+                    deltaTransform.position += endClipSample.position - previousClipSample.position;
```
Comment (Dreaming381) [inline Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs]: I know you are sick of me discussing this, and I am willing to let this through as you have spent more time testing this, but I don't think the math is correct here.

Imagine we have an animation for 1 second duration that moves the object on the x axis from 0 to 1. Now imagine that our speedModifiedDeltaTime is 0.01 and is equivalent to deltaTime and our current time in the clip is 0.005 which means we looped. Naturally, the expected root motion delta should also be 0.01.

We sample deltaTransform and get a value of 0.005. For previousClipSample, we sample at -0.005 which ACL clamps to 0 (ACL clamps all time values from 0 to the clip duration), so we get a value of 0 there. And for endClipSample, we sample at clip duration and get the value of 1.

Thus our final delta = 0.005 + 1 - 0 = 1.005 which does not equal 0.01.

File: Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs [hunk 4]
```diff
@@ -458,11 +459,172 @@ internal static bool Approximately(float a, in float b)
...
+                    deltaTransform.rotation = math.mul(deltaTransform.rotation, math.mul(math.inverse(endClipSample.rotation), previousClipSample.rotation));
+                }
```
Comment (Dreaming381) [inline Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs]: Again, not a big deal if you don't want to support this yet, but you don't have scale and stretch factored in here. They use the same math. They function like position, except every addition is multiplication and every subtraction is division.

**Review by Dreaming381 — COMMENTED** [review]
With the exception of MeshDeformDataSmartBlobberSystem.cs (just remove that change from the PR as I will fix it in my own commit), I am willing to merge the changes. I will leave that up to you to decide.

One other thing I plan to change is to rename `Latios.Mimic.Mecanim` to `Latios.Mimic.Addons.Mecanim`. You are free to make that change too if you like.

File: Packages/com.latios.latiosframework/Kinemation/Authoring/BakingSystems/MeshDeformDataSmartBlobberSystem.cs (hunk 1 above)
Comment (Sovogal) [reply Packages/com.latios.latiosframework/Kinemation/Authoring/BakingSystems/MeshDeformDataSmartBlobberSystem.cs]: It seemed to be working just fine, but I wasn't trying to test any edge cases, nor did I know what those could be.  I'll undo the change and let you correct it.

File: Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs (hunk 4 above)
Comment (Sovogal) [reply Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs]: I will transform the other values.  Completely forgot our conversation about QVVS.

File: Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs [hunk 5]
```diff
@@ -486,32 +486,42 @@ internal static bool Approximately(float a, in float b)
...
                     deltaTransform.position += currentClipSample.position - previousClipSample.position;
                     deltaTransform.rotation = math.mul(currentClipSample.rotation, math.inverse(previousClipSample.rotation));
+                    deltaTransform.scale *= currentClipSample.scale / previousClipSample.scale;
+                    deltaTransform.stretch *= currentClipSample.stretch / previousClipSample.stretch;
                 }
 
                 currentRoot.position += deltaTransform.position * blendWeight;
                 currentRoot.rotation = math.slerp(currentRoot.rotation, math.mul(currentRoot.rotation, deltaTransform.rotation), blendWeight);
+                currentRoot.scale *= deltaTransform.scale * blendWeight;
+                currentRoot.stretch *= deltaTransform.stretch * blendWeight;
```
Comment (Dreaming381) [inline Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs:525]: These still need to be additive with the currentRoot values starting at zero rather than identity. Otherwise the blendWeights scale everything down.

File: Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs [hunk 6]
```diff
@@ -565,16 +579,22 @@ internal static bool Approximately(float a, in float b)
...
+                previousRoot.scale *= sampleTransform.scale * blendWeight;
+                previousRoot.stretch *= sampleTransform.stretch * blendWeight;
             }
 
             var rootDelta = TransformQvvs.identity;
 
             rootDelta.position = currentRoot.position - previousRoot.position;
             rootDelta.rotation = math.mul(currentRoot.rotation, math.inverse(previousRoot.rotation));
+            rootDelta.scale = currentRoot.scale / previousRoot.scale;
+            rootDelta.stretch = currentRoot.stretch / previousRoot.stretch;
```
Comment (Dreaming381) [inline Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs]: I think you have this backwards. currentRoot represents the root motion accumulated after the transition, whereas previousRoot represents the root motion accumulated before the hard exit time transition (rare). I believe we would want to add these, not subtract them.

File: Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs (hunk 5 above)
Comment (Sovogal) [reply Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs:525]: I see what you're saying.  Will change and review other scale and stretch assignments.

File: Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs (hunk 6 above)
Comment (Sovogal) [reply Packages/com.latios.latiosframework/Mimic/Mecanim/Internal/MecanimInternalUtilities.cs]: I don't think that's quite right.  Let's say the current root is 1.2f scale, and previous root is 1.2f scale.  If we multiplied them, that would make the result 1.44f, which would grow the transform instead of keeping it the same.  However, looking at the optimized skeleton system, I see I'm not multiplying the existing root by the delta, as I am in the exposed bone system.  I'll run a few more tests against both types of skeletons before I adjust it.

**Comment by Sovogal** [comment]
Closed to resync with working 0.9 branch
